### ISSUE TITLE
Initiating connection with ranch_tcp and upgrading connection (tcp socket) to ranch_ssl

### ISSUE THREAD

**Issue by everydaysasaturday** [ninenines/ranch#139]
Is it possible to accept connections with ranch_tcp and then upgrading those connections to ranch_ssl? 

So like ssl-module supports:
start with tcp -> accept -> tcp connection established -> upgrade connection to ssl -> ssl_accept, get new ssl socket -> controlling_process call for new socket -> ssl connection established

Or would this be a new feature?

**Comment by essen** [comment]
It's currently not possible. I'm open for it, I think that's one of the longest living todo Ranch has.

**Comment by essen** [comment]
You can of course call ssl:ssl_accept directly and it'll work, it's just not possible through the ranch API.

**Comment by everydaysasaturday** [comment]
Thanks for a quick answer!

That's true, it works nicely with ssl:ssl_accept directly.
The diamond thing here would be that there would be a module like ranch_tcp2ssl that would do it behind the curtains. It would be super-sweet! 

Is there any plans of adding it to Ranch in the near future?

Ranch rocks!

**Comment by essen** [comment]
Ideally it should be an accept_ack function with SSL options added and returning a new socket. Perhaps in Ranch 2 we should make all accept_ack functions return the new socket, since we can't use the socket before that anyway. Then we could have two similar accept_ack functions, just one that does an upgrade to SSL. Food for thoughts for now.

**Comment by everydaysasaturday** [comment]
Actually that is waay more feasible and streamlined than what I thought!
Sounds promising and looking forward for this one!

**Comment by juhlig** [comment]
I have been thinking about this for a while, since I'm just in need of the same thing. Upgrading "by hand" works well, but it would be nice to do that via ranch.

Changing `accept_ack` to return a new socket instead of `ok` appears not feasible to me. It will probably break existing code, especially so since [the user guide on protocols](https://ninenines.eu/docs/en/ranch/1.4/guide/protocols/) up to 1.4 has `ok=ranch:accept_ack(Ref)` in it, so many people will have that in their code (at least I have it in mine everywhere I use ranch).

How about an (optional?) new callback like `transition/5` instead? (I can't think of a better name right now. `upgrade` however seems too one-sided since with custom protocols, it needs not neccesarily be any sort of _upgrade_ ) Given `Ref`, `Transport`, `Socket`, `Options` (?) and some sort of `TransitionOptions`, it could transit the given socket to a socket of another kind, in a way that the transport it is called on knows how to do. It would then return new `Ref` (usually identical to the old one), `Transport`, `Socket` and `Options` fitting the transport the socket has been transited to, or an error (or crash?) if the given socket cannot be transited from the given transport.

Eg, `transit/5` in `ranch_ssl` might take a socket of transport `ranch_tcp`, with the needed ssl options like certificate etc in the `TransitOptions`, and return `ranch_ssl` as the new transport and the ssl socket as the new socket. For transports other than `ranch_tcp`, it will return an error. Roughly like this:
```erlang
transition(Ref, ranch_tcp, Socket, Opts, TransitionOpts) ->
    % ssl_accept etc
    {ok, {Ref, ranch_ssl, NewSock, NewOpts}};

transition(_, _, _, _, _) ->
    {error, not_supported}.
```

I checked `ranch_conns_sup`, and the `Transport` it carries around seems to be needed only for connection purposes? When you're upgrading "by hand", you have no say there, either, anyway.

**Comment by juhlig** [comment]
On second thought, `Ref` and `Opts` are probably not needed or even useful, so a `transition/3` should suffice.

**Comment by juhlig** [comment]
On third thought, a timeout should be in it, so that makes it `transition/4` =^^=

**Comment by essen** [comment]
Yep that might be a better choice. I should look into implementing HTTP TLS Upgrade and see how it all fits.

**Comment by juhlig** [comment]
I could take a stab at it maybe next week, if you like

**Comment by juhlig** [comment]
I tried my hand on it today (yes, I said next week, but there were long wait times on other projects, so...), and I stumbled upon the following:

* making the `transition` function a _compulsory_ `ranch_transport` callback will break existing custom transports in that it will generate a warning if it isn't implemented, and with `warnings_as_errors`, a custom transport will not compile any longer
* making it an _optional_ callback makes no sense, because transport functions have to be called like `Transport:function(Args)`, and not via the `ranch_transport` module which could fall back on a default implementation. So if you would want to use `transition`, you would have to know if a transport implements it or not
* assuming further that you would have to specifically call `transition` on the transport module you want to transition to (eg, `ranch_ssl:transition(...)`), leads to the conclusion that a function like `transition` might as well be a transport-specific function instead of a generic callback

So, my proposition is to not make `transition`a callback of `ranch_transport` but to implement it as an independent, transport-specific function in `ranch_ssl` only, in which case it could as well be named `upgrade` or `upgrade_from` or something. No ambiguities, no breaking existing custom transports.

Custom transports can implement their own functions for transitioning as they like (and name them as they like). In fact, they can do that right now, anyway. But, when you're using a custom transport you will already have to know certain things about them (ie, how they behave, and what they are there for in the first place). Most of the time your own applications will be the sole users of your custom transports, too.

What do you think?

**Comment by essen** [comment]
Well that's pretty much why it has not been implemented yet, it all boils down to just calling the ssl function and there's no real use in wrapping it in ranch_ssl if it's only going to be there and not have an equivalent in ranch_tcp or other. But otherwise yes, `upgrade` or `upgrade_from` sounds better, and I wouldn't be too worried about warnings as errors in general.

**Comment by juhlig** [comment]
Yeah, I guessed.

One thing a ranch wrapper could do on top of simply calling the ssl functions is check options and provide reasonable defaults, like in the `listen` and subsequent `do_listen` functions.

Apart from that, having a function like this in ranch eliminates that nasty "I'm doing something hacky" feeling you get when upgrading by hand. You have to read some of ranch's code to see that it is actually ok to do it this way. Still, it is not guaranteed to be ok forever, so the feeling remains.

So, anyway, how would you like me to approach this?

**Comment by essen** [comment]
Those pesky feelings, why do people have those anyway?

OK then sounds good. I think a `ranch_ssl:upgrade/3,4` would do like you describe it, with validation and an error returned if a parameter is incorrect. Leave it optional and we can see later if it makes sense to add to ranch_transport. Most importantly let's try to see it in action, I'll try to make time to do the TLS upgrade to see if it all makes sense.

**Comment by juhlig** [comment]
I saw a book titled "F*ck Feelings" at a store recently, maybe I should read it ;)

I'll work on it next week then

**Comment by essen** [comment]
Sounds like a good book. Though if you're going for that kind of book, 12 Rules for Life is by far the most important to read.

**Comment by juhlig** [comment]
I'll check it out :)

That was published last month, and you read it already?

**Comment by essen** [comment]
I should receive it next week but I already know most of what it contains. I'm listening to the guy a lot.

**Comment by juhlig** [comment]
I must admit that I never heard of him before, but now I'm interested :)

As for SSL upgrading, I have a first draft (PR #180). Early again, yes =^^=

**Comment by essen** [comment]
Yeah. I'll see if I can make time but don't keep your hopes up.

**Comment by juhlig** [comment]
No worries ;)

**Comment by essen** [comment]
I think the new `Transport:handshake/3` function allows doing this at any time since `ssl:accept_ack` and `ssl:handshake` accepts both TCP and TLS sockets. Obtain the TCP socket from `ranch:handshake/1,2` and then call Transport:handshake as necessary. It's been documented so I consider this resolved and it will be available in Ranch 1.6 onward.

What's missing is the ability to downgrade a connection from TLS to TCP but as far as I know it's not possible with the current `ssl` interface.
